test/filters/test_filters.cpp

```diff
@@ -1326,6 +1326,105 @@ TEST (VoxelGridCovariance, Filters)
+TEST (VoxelGridMinPoints, Filters)
+{
+  // Setup input with 4 clusters, single point at 0,0,0 and 1,1,1 with 5 point cluster around 0.11,0.11,0.11 and 6 point cluster around 0.31,0.31,0.31
+  PointCloud<PointXYZRGB>::Ptr input(new PointCloud<PointXYZRGB>());
+
+  input->emplace_back(0.0f, 0.0f, 0.0f);
+  std::vector<float> offsets {0.001, 0.002, 0.003, -0.001, -0.002, -0.003};
+  for (unsigned int i=0; i<5; ++i) {
+    input->emplace_back(0.11f+offsets[i], 0.11f+offsets[i], 0.11f+offsets[i],200,0,0);
+    input->emplace_back(0.31f+offsets[i], 0.31f+offsets[i], 0.31f+offsets[i],0,127,127);
+  }
+  input->emplace_back(0.31f+offsets[5], 0.31f+offsets[5], 0.31f+offsets[5],0,127,127);
+  input->emplace_back(1.0f, 1.0f, 1.0f);
+
+  // Test the PointCloud<PointT> VoxelGrid filter method
+  PointCloud<PointXYZRGB> output;
+  VoxelGrid<PointXYZRGB> grid;
+  // Run filter on input with MinimumPoints threshold at 4
+  grid.setLeafSize (0.02f, 0.02f, 0.02f);
+  grid.setInputCloud (input);
+  grid.setMinimumPointsNumberPerVoxel(4);
+  grid.setDownsampleAllData(true);
+  grid.filter (output);
+
+  // Verify 2 clusters (0.11 and 0.31) passed threshold and verify their location and color
+  EXPECT_EQ (output.size (), 2);
+
+  EXPECT_NEAR (output[0].x, 0.11, 1e-2);
+  EXPECT_NEAR (output[0].y, 0.11, 1e-2);
+  EXPECT_NEAR (output[0].z, 0.11, 1e-2);
+  EXPECT_NEAR (output[0].r, 200, 1);
+
+  EXPECT_NEAR (output[1].x, 0.31, 1e-2);
+  EXPECT_NEAR (output[1].y, 0.31, 1e-2);
+  EXPECT_NEAR (output[1].z, 0.31, 1e-2);
+  EXPECT_NEAR (output[1].g, 127, 1);
+  EXPECT_NEAR (output[1].b, 127, 1);
+
+  // Run filter again on input with MinimumPoints threshold at 6
+  grid.setMinimumPointsNumberPerVoxel(6);
+  grid.setDownsampleAllData(false);
+  grid.filter (output);
+
+  // Verify 1 cluster (0.31) passed threshold and verify location
+  EXPECT_EQ (output.size (), 1);
+
+  EXPECT_NEAR (output[0].x, 0.31, 1e-2);
+  EXPECT_NEAR (output[0].y, 0.31, 1e-2);
+  EXPECT_NEAR (output[0].z, 0.31, 1e-2);
+
+  // Test the pcl::PCLPointCloud2 VoxelGrid filter method
+  PCLPointCloud2 output_pc2;
+  VoxelGrid<PCLPointCloud2> grid_pc2;
+  PCLPointCloud2::Ptr input_pc2(new PCLPointCloud2());
+
+  // Use same input as above converted to PCLPointCloud2
+  toPCLPointCloud2(*input, *input_pc2);
+
+  // Run filter on input with MinimumPoints threshold at 4
+  grid_pc2.setLeafSize (0.02f, 0.02f, 0.02f);
+  grid_pc2.setInputCloud (input_pc2);
+  grid_pc2.setMinimumPointsNumberPerVoxel(4);
+  grid_pc2.setDownsampleAllData(true);
+  grid_pc2.filter (output_pc2);
+
+  // Convert back to PointXYZRGB for easier data access
+  PointCloud<pcl::PointXYZRGB>::Ptr out_pc( new pcl::PointCloud<pcl::PointXYZRGB> );
+  pcl::fromPCLPointCloud2( output_pc2, *out_pc );
+
+  // Verify 2 clusters (0.11 and 0.31) passed threshold and verify their location and color
+  EXPECT_EQ (out_pc->points.size (), 2);
+
+  EXPECT_NEAR (out_pc->at(0).x, 0.11, 1e-2);
```

Comment:
Can we instead compare the points in PointCloud with PCLPointCloud2 and expect eq on them?

Reply:
They aren't quite equal because of the centroid averaging and my offsets not summing to a particular input point, but I set them to be near a point in the expected input cluster